SU2_CFD/src/solver_direct_scalar.cpp

```diff
@@ -0,0 +1,1716 @@
+  
+  SetImplicitPeriodic(true);
+  
+}
+
+CScalarSolver::~CScalarSolver(void) {
+  
+  if (Inlet_ScalarVars != NULL) {
+    for (unsigned long iMarker = 0; iMarker < nMarker; iMarker++) {
+      if (Inlet_ScalarVars[iMarker] != NULL) {
+        for (unsigned long iVertex = 0; iVertex < nVertex[iMarker]; iVertex++) {
+          delete [] Inlet_ScalarVars[iMarker][iVertex];
+        }
+        delete [] Inlet_ScalarVars[iMarker];
+      }
+    }
+    delete [] Inlet_ScalarVars;
+  }
+  
+  if (FlowPrimVar_i != NULL) delete [] FlowPrimVar_i;
+  if (FlowPrimVar_j != NULL) delete [] FlowPrimVar_j;
+  if (lowerlimit != NULL)    delete [] lowerlimit;
+  if (upperlimit != NULL)    delete [] upperlimit;
+  if (nVertex != NULL)       delete [] nVertex;
+  if (Scalar_Inf != NULL)    delete [] Scalar_Inf;
+
+}
+
+void CScalarSolver::Upwind_Residual(CGeometry *geometry,
+                                    CSolver **solver_container,
+                                    CNumerics *numerics,
+                                    CConfig *config,
+                                    unsigned short iMesh) {
+  
+  su2double *Scalar_i, *Scalar_j, *Limiter_i = NULL, *Limiter_j = NULL;
+  su2double *V_i, *V_j, **Gradient_i, **Gradient_j;
+  su2double Project_Grad_i, Project_Grad_j;
+  
+  unsigned long iEdge, iPoint, jPoint;
+  unsigned short iDim, iVar;
+  
+  bool muscl         = (config->GetMUSCL_Scalar() && (iMesh == MESH_0));
+  bool limiter       = (config->GetKind_SlopeLimit_Scalar() != NO_LIMITER);
+  bool grid_movement = config->GetGrid_Movement();
+  
+  for (iEdge = 0; iEdge < geometry->GetnEdge(); iEdge++) {
+    
+    /*--- Points in edge and normal vectors ---*/
+    
+    iPoint = geometry->edge[iEdge]->GetNode(0);
+    jPoint = geometry->edge[iEdge]->GetNode(1);
+    numerics->SetNormal(geometry->edge[iEdge]->GetNormal());
+    
+    /*--- Primitive variables w/o reconstruction ---*/
+    
+    V_i = solver_container[FLOW_SOL]->node[iPoint]->GetPrimitive();
+    V_j = solver_container[FLOW_SOL]->node[jPoint]->GetPrimitive();
+    numerics->SetPrimitive(V_i, V_j);
+    
+    /*--- Scalar variables w/o reconstruction ---*/
+    
+    Scalar_i = node[iPoint]->GetSolution();
+    Scalar_j = node[jPoint]->GetSolution();
+    numerics->SetScalarVar(Scalar_i, Scalar_j);
+    
+    /*--- Grid Movement ---*/
+    
+    if (grid_movement)
+      numerics->SetGridVel(geometry->node[iPoint]->GetGridVel(), geometry->node[jPoint]->GetGridVel());
+    
+    if (muscl) {
+      
+      for (iDim = 0; iDim < nDim; iDim++) {
+        Vector_i[iDim] = 0.5*(geometry->node[jPoint]->GetCoord(iDim) - geometry->node[iPoint]->GetCoord(iDim));
+        Vector_j[iDim] = 0.5*(geometry->node[iPoint]->GetCoord(iDim) - geometry->node[jPoint]->GetCoord(iDim));
+      }
+      
+      /*--- Mean flow primitive variables using gradient reconstruction and limiters ---*/
+      
+      Gradient_i = solver_container[FLOW_SOL]->node[iPoint]->GetGradient_Primitive();
```

Comment:
This is what the turbulent solver does almost verbatim. Are you moving the heat and turbulence solvers to this framework?

Reply:
Great, it would be nice to re-use the muscl reconstruction bit of the code done in UpwindResidual too.

Reply:
I want to change my mind, can you keep the solvers as is for now? (merging #753 is triple the work otherwise...)

Reply:
Should I avoid just the solvers? Does it cause an issue to adjust the viscous numerics? 

I would like to clean up some things for this PR and hook up the new output classes for the scalars too if that won't cause any problems for you either.

Reply:
Numerics is ok to change, selfish request but if the turbulence and heat solvers are changed significantly, I would have to re-apply the changes to them, unless you do that work already in the framework of #753 :p

Reply:
I have now updated the classes so that all of the turbulent convective, viscous, and variable classes inherit from a scalar base class versions. This removes some duplicate code. I have not yet adjusted the solver classes, since this is a little more delicate.

Btw, this PR has now also been updated with the new contiguous variable structures and the new output.